src/dmd/identifier.d

```diff
@@ -153,23 +166,75 @@ nothrow:
      */
     extern (D) static Identifier generateIdWithLoc(string prefix, const ref Loc loc)
     {
-        OutBuffer buf;
-        buf.writestring(prefix);
-        buf.writestring("_L");
-        buf.print(loc.linnum);
-        buf.writestring("_C");
-        buf.print(loc.charnum);
-        auto basesize = buf.peekSlice().length;
-        uint counter = 1;
-        while (stringtable.lookup(buf.peekSlice().ptr, buf.peekSlice().length) !is null)
+        import dmd.root.filename: absPathThen;
+
+        // see below for why we use absPathThen
+        return loc.filename.absPathThen!((absPath)
         {
-            // Strip the extra suffix
-            buf.setsize(basesize);
-            // Add new suffix with increased counter
-            buf.writestring("_");
-            buf.print(counter++);
-        }
-        return idPool(buf.peekSlice());
+
+            // this block generates the "regular" identifier, i.e. if there are no collisions
+            OutBuffer idBuf;
+            idBuf.writestring(prefix);
+            idBuf.writestring("_L");
+            idBuf.print(loc.linnum);
+            idBuf.writestring("_C");
+            idBuf.print(loc.charnum);
+
+            // This block generates an identifier that is prefixed by the absolute path of the file
+            // being compiled. The reason this is necessary is that we want unique identifiers per
+            // module, but the identifiers are generated before the module information is available.
+            // To guarantee that each generated identifier is unique without modules, we make them
+            // unique to each absolute file path. This also makes it consistent even if the files
+            // are compiled separately. See issues:
+            // https://issues.dlang.org/show_bug.cgi?id=16995
+            // https://issues.dlang.org/show_bug.cgi?id=18097
+            // https://issues.dlang.org/show_bug.cgi?id=18111
+            // https://issues.dlang.org/show_bug.cgi?id=18880
+            // https://issues.dlang.org/show_bug.cgi?id=18868
+            // https://issues.dlang.org/show_bug.cgi?id=19058.
+            OutBuffer fullPathIdBuf;
+
+            if (absPath)
+            {
+                // replace characters that demangle can't handle
+                for (auto ptr = absPath; *ptr != '\0'; ++ptr)
+                {
+                    // see dmd.dmangle.isValidMangling
+                    // Unfortunately importing it leads to either build failures or cyclic dependencies
+                    // between modules.
+                    if (*ptr == '/' || *ptr == '\\' || *ptr == '.' || *ptr == '?' || *ptr == ':')
+                        *ptr = '_';
+                }
+
+                fullPathIdBuf.writestring(absPath);
+                fullPathIdBuf.writestring("_");
+            }
+
+            fullPathIdBuf.writestring(idBuf.peekSlice());
+            const fullPathIdLength = fullPathIdBuf.peekSlice().length;
+            uint counter = 1;
```

Comment:
I think the counter/StringTable could be simplified and possibly accelerated/requiring less memory via `static uint[string][Loc] counters`, indexed by `counters[prefix][loc]`; checking if there's an existing entry and incrementing/setting it to 1 when done. I might give that a try as a follow-up.